src/main/host/syscall/handler/mod.rs

```diff
@@ -214,6 +215,30 @@ pub fn read_sockaddr(
     Ok(Some(addr))
 }
 
+/// Writes `val` to `val_ptr`, but will only write a partial value if `val_len` is smaller than the
+/// size of `val`. Returns the number of bytes written.
+///
+/// The generic type must be given explicitly to prevent accidentally writing the wrong type.
+///
+/// ```ignore
+/// let bytes_written = write_partial::<i32, _>(mem, foo(), ptr, len)?;
+/// ```
+pub fn write_partial<U: NoTypeInference<This = T>, T: pod::Pod>(
+    mem: &mut MemoryManager,
+    val: &T,
+    val_ptr: PluginPtr,
+    val_len: usize,
+) -> Result<usize, SyscallError> {
+    let val_len = std::cmp::min(val_len, std::mem::size_of_val(val));
+
+    let val = &pod::as_u8_slice(val)[..val_len];
+    let val_ptr = TypedPluginPtr::new::<MaybeUninit<u8>>(val_ptr, val_len);
+
+    mem.copy_to_ptr(val_ptr, val)?;
```

Comment:
Hmm.... orthogonal to this PR, but I think we have a safety hazard in the `Pod` module. Since `MaybeUninit<T>` is declared as `Pod` where `T` is `Pod`, and we allow conversion from any `Pod` to `u8` slices, we have a path to go from uninitialized data to a regular `u8` slice without using `unsafe`.

I don't *think* we'll actually get unsound behavior in this specific instance since `copy_to_ptr` doesn't touch the bytes in userspace anyway.

Reply:
The `as_u8_slice` function actually returns a `&[MaybeUninit<u8>]`, which we don't cast to a `&[u8]` slice in the memory manager before passing to nix, so I think we're okay here?